### PR TITLE
LPS-121920

### PR DESCRIPTION
**Pull request by jonmak08**
Hey @liferay-echo,

Attached is an update for http://issues.liferay.com/browse/LPS-121920.

This is a followup for some minor changes that were already pushed up with https://github.com/liferay-echo/liferay-portal/pull/3575.

**Test Plan:**

1. Add **web content** to Liferay
2. Delete web content to Recycle Bin
3. Delete/Restore content in Recycle Bin

106336446-c5745b00-6243-11eb-825b-2813f458f43a

Please let me know if you have any questions. Thanks!

### CODE REVIEW

**Comment by ealonso** [comment]
ci:test:sf

**Comment by ealonso** [comment]
ci:test:relevant

**Comment by victorg1991** [comment]
Taking a look :)

File: modules/apps/trash/trash-web/src/main/resources/META-INF/resources/js/TrashManagementToolbarPropsTransformer.js [hunk 1]
```diff
@@ -23,18 +23,30 @@ export default function propsTransformer({
 				Liferay.Language.get('are-you-sure-you-want-to-delete-this')
 			)
 		) {
-			submitForm(document[`${portletNamespace}fm`]);
+			const form = document.getElementById(`${portletNamespace}fm`);
+
+			if (!form) {
+				return;
```
Comment (victorg1991) [inline modules/apps/trash/trash-web/src/main/resources/META-INF/resources/js/TrashManagementToolbarPropsTransformer.js:29]: I think we may need some logging here, so in case someone changes the jsp and remove/Rename the form It will be warned that something went wrong instead of failing silently, makes sense?

We can use a simple console.error:
```
if (process.env.NODE_ENV === 'development') {
      console.error(error);
}
```

or just throw an exception here like 

`throw new Error('form with id: ${formId} can  not be found')`

Comment (markocikos) [reply modules/apps/trash/trash-web/src/main/resources/META-INF/resources/js/TrashManagementToolbarPropsTransformer.js:29]: @victorg1991 We have a convention to followup each DOM query with a null-check. In it, we don't throw error or add logs. You can see this in many thousands of places throughout portal, just search for `querySelector` or `getElementById`.

If the goal would be only to prevent developer mistake, I agree with you that falling silently doesn't make sense. For this, it is better to let it fail with default error. We add logs such as [this one](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/modal/Modal.js#L82-L87) to help developer if the mistake is not obvious and manifests in an obscure way. In the linked example, it is the case that modal should have only one form. Wrong `id`, in my opinion, is an obvious mistake.

But, the purpose of the null-check after DOM query is not to prevent a developer mistake. The main purpose is to cover weird edge cases, caused by user actions during temporary lifecycles of components. Code can unexpectedly execute during async resource loading, navigation etc. The secondary purpose is to cover workflows that remove parts of DOM, while references still remain in event handlers. DOM mutations are very common and it's impossible to predict all interactions. In these cases, it is better to do gracefully do nothing than throw error in JS that can disrupt other JS code.

@wincent please weigh in. Did I miss something? Do these conventions still make sense?

Comment (markocikos) [reply modules/apps/trash/trash-web/src/main/resources/META-INF/resources/js/TrashManagementToolbarPropsTransformer.js:29]: /cc @julien @ambrinchaudhary

**Review by markocikos — APPROVED** [review]

File: modules/apps/trash/trash-web/src/main/resources/META-INF/resources/js/TrashManagementToolbarPropsTransformer.js (hunk 1 above)
Comment (julien) [reply modules/apps/trash/trash-web/src/main/resources/META-INF/resources/js/TrashManagementToolbarPropsTransformer.js:29]: > I think we may need some logging here, so in case someone changes the jsp and remove/Rename the form It will be warned that something went wrong instead of failing silently, makes sense?

I agree with what @markocikos said.

If the developer removes the form/changes the JSP (like we all do) he/she should check that his/her changes don't break anything.

And if we don't want to fail silently, we could remove 

```js
if (!form) {
   return;
}
```

Although, we've decided that we would check that.

Comment (victorg1991) [reply modules/apps/trash/trash-web/src/main/resources/META-INF/resources/js/TrashManagementToolbarPropsTransformer.js:29]: If the form is not there it will fail silently which will result in a really bad user experience, right? We may even show a generic toast to let the user know that something happened, but just fail silently doesn't seem like the best thing to do 🤔 what do you think?

I thought about the log because the only possible case I though of the form not being there is that someone refactored/changed that part and forget about the form or change the name

Comment (markocikos) [reply modules/apps/trash/trash-web/src/main/resources/META-INF/resources/js/TrashManagementToolbarPropsTransformer.js:29]: > If the form is not there it will fail silently which will result in a really bad user experience, right? We may even show a generic toast to let the user know that something happened, but just fail silently doesn't seem like the best thing to do thinking what do you think?

If there is a legitimate use case where the form is not present, there should definitely be a toast or something similar in UI. I am assuming here that the form should always be present, and the reason why it is not is either a developer mistake or an unexpected weird edge case. In most cases it's better not to check for developer mistake. Check after DOM query goes against this, because we have other considerations, explained in https://github.com/liferay-echo/liferay-portal/pull/3681#discussion_r572814761. The convention is a compromise.

Comment (victorg1991) [reply modules/apps/trash/trash-web/src/main/resources/META-INF/resources/js/TrashManagementToolbarPropsTransformer.js:29]: Yeah, yeah, I understand the consideration, but if there isn't a legitimate use case where the form is not present, maybe we shouldn't be adding the `if` in the first place.

I think that there are places where this kind of defensive checks makes sense, but in this particular case that we are executing and action triggered by the user we should do the action or warn the user that the action hasn't been performed. In some cases, like this one, we may not know the cause of the failure, but I believe we should warn the user anyway.

But all in all this it is not a big deal, just a minor suggestion trying to improve the overall user experience :)

Comment (markocikos) [reply modules/apps/trash/trash-web/src/main/resources/META-INF/resources/js/TrashManagementToolbarPropsTransformer.js:29]: It would be a huge deal to decide to consistently provide UI messages for DOM query failures. Very costly, just to cover edge cases. As I mentioned, there are thousands of places with DOM queries. A null-check is a simple "good enough" compromise. An ideal solution is to rewrite entire module frontend to React, which I highly recommend! There will be no issues with DOM queries after that :slightly_smiling_face:

Comment (ealonso) [reply modules/apps/trash/trash-web/src/main/resources/META-INF/resources/js/TrashManagementToolbarPropsTransformer.js:29]: Hey @markocikos,
Really good idea, Could you please send us a pr with the rewrite of module to React? We are going to really appreciate it.
Thanks!!

Comment (markocikos) [reply modules/apps/trash/trash-web/src/main/resources/META-INF/resources/js/TrashManagementToolbarPropsTransformer.js:29]: Of course, I was not serious about rewriting entire module just to fix the DOM issue. It is a worthy goal to reach one day, but I am well aware you have other priorities. The whole epic of changing tags to Clay 3 is a form of compromise, where we are introducing highest amount of React with the least impact to you, to existing module structure.

Bottom line is, we have a convention and recommendation that there should be a null-check after each DOM query. @victorg1991 if you adamantly disagree, feel free to close this PR. If you wish to escalate initiative to change it, we can. I am personally for keeping the convention, as explained in comments above. Even if I wasn't, I think it will almost certainly be a futile effort.

If you wish to add a UI note that informs user of the failure, you are free to do so. We typically don't do this, as it is considered too much of an effort for small benefit. In any case, it is outside of scope of this ticket.

Comment (wincent) [reply modules/apps/trash/trash-web/src/main/resources/META-INF/resources/js/TrashManagementToolbarPropsTransformer.js:29]: > @wincent please weigh in. Did I miss something? Do these conventions still make sense?

Seeing as I have been summoned, will [link to our recommendations here](https://github.com/liferay/liferay-frontend-projects/blob/master/guidelines/dxp/debugging.md#on-the-use-of-consolelog). Now, there is a lot of nuance that we could probably add to that document, but the core of it is the following (emphasis added for the most important bit):

> Another problem with extensive logging is that it can easily be perceived as "console spew", and that reflects badly on the quality of the product. Customers may not have the necessary context to distinguish messages that indicate real problems from ones which merely seek to helpfully inform. Logging may actually scare or alarm people more than it helps them, especially when you consider the population of users — often customers of our customers — who know enough to open a browser console but who know little or nothing about the underlying system.
>
> `console.error` and `console.warn` may be used sparingly to communicate actual problems, but in keeping with our philosophy, **always prefer to implement graceful recovery and user experiences that are designed with resilience in mind**. The section on NODE_ENV=development below describes one way in which you can avoid the "console spew" problem but still log useful information in a development context while keeping production environments pristine.

So that's the thing on which @markocikos was basing his original advice:

- We pretty much _never_ want to `throw` an error (unless it is going to be caught by some layer higher up).
- We mostly want error logging to be gated by a `NODE_ENV` check (like it says above "use sparingly").

Both of those boil down to the the "golden rule" that we never want a customer to see something that looks like a programming error. By that standard, even a silently "graceful" failure is better than a thrown or logged error. To take it to the next level, you do what @victorg1991 said and either implement a graceful recovery workflow (eg. a toast with a retry button), or you try to make the UI at least continue working in a degraded way despite the error.

As we move towards TypeScript there will be a lot of places where we need these kinds of "`null` checks" because, technically, `getElementById()` _can_ always return `null` even if you "know" that the element has to be there (at the level of the type system, there are no such guarantees). So in those cases, we're forced to implement an `if (!form) { return }` kind of check to keep the type system happy, and degrade gracefully (ie. no thrown errors) in the event that the unthinkable happens and the element isn't there. This is perfectly fine IMO and a probably-never-actually-executed-early-`return` is a reasonable price to pay for all the other good things we'll get with the type system.

In this case, we also have the hypothetical scenario that "someone changes the jsp and remove/Rename the form". I don't think there's any real value to logging (or showing toasts etc) to accommodate this scenario. If developers doing manual testing, CI runs, and QA don't catch this kind of regression (by noticing that the component isn't rendering properly) then we're not going to notice the much more subtle signal of a message in the console or a toast.

So, as I said above, a lot of nuance to think about and it all depends very much on context. On the Frontend Infrastructure, we can provide guidelines and recommendations for this code, but it's ultimately up to the product team that owns the functionality to decide how far they go down the road of "graceful recovery" and so on. In contrast, the kind of stuff that is black-and-white and which we have hard "rules" for are things like "don't ever return a promise eg. from a `fetch` without a `catch` somewhere upstream.

tl;dr: I'd go with `if (!form) { return; }` to make the eventual migration to TypeScript easier, but it's fine if the product team wants to do something else.

Comment (victorg1991) [reply modules/apps/trash/trash-web/src/main/resources/META-INF/resources/js/TrashManagementToolbarPropsTransformer.js:29]: Thanks @wincent for the thorough explanation. I didn't think about typescript at all, without taking that into account that if hadn't made a lot of sense to me. So having that in mind, don't delay this more and let's get this merge. We'll have this things into account in future PRs

Thanks again :)

**Comment by victorg1991** [comment]
ci:forward

**Comment by victorg1991** [comment]
@jonmak08 Thanks for the pr and sorry for all the noise!

**Comment by jonmak08** [comment]
Thanks for all the input and feedback guys!
